### PR TITLE
Add support for exposing server certificate

### PR DESCRIPTION
**Pull request by Nicoretti**
Dear pytest-localserver maintainers,

First off, I want to send a big :+1: your way. Thanks for providing and maintaining this plugin!

While I was using it in the context of HTTPS, I stumbled upon a few minor issues.
I tried addressing all of these in this PR.

I hope you find the updates useful. If there's anything that needs tweaking or if you'd like me to make any changes, please don't hesitate to let me know.

best

Nico

### CODE REVIEW

**Comment by diazona** [comment]
Thanks for sending a contribution Nico! I'll review this properly later (over the weekend, or tonight if I have time), but for starters, I wonder if you could explain the use cases that led you to propose these changes? (This will probably be good info to edit into the commit messages at some point, but for now we can discuss it in comments.)

**Comment by Nicoretti** [comment]
Hi @diazona, 

No worries take your time. My use cases are quite similar to the ones I've covered in the unit tests of this PR.

When dealing with `SSL/TLS`, there are three interesting variations, especially concerning server certificate validation when it's not signed by an official Certificate Authority (CA).

1. Verify and abort the connection if validation fails.
2. Proceed without verification (implicitly trust the certificate).
3. Add the certificate to a list of trusted certificates.

Testing scenario `3.` requires access to the server's certificate. The `httpsserver` fixture, with its self-signed certificate, is ideal for this purpose. However, to accurately test this scenario, the server certificate must be made accessible to the client, which was not previously implemented.

While I was on it, I encountered and addressed a few other issues:

- The parameters for `sslcontext` were passed in an incorrect order. This mistake was subtle, as it didn't manifest while the server's key and certificate were combined in a single `.pem` file.
- I separated the `server.pem` into individual key and certificate files (`key` and `crt`), allowing them to be exposed separately for more granular control.

best
Nico

**Comment by Nicoretti** [comment]
For more context: My use cases can be found [here](https://github.com/exasol/bucketfs-python/blob/main/exasol/bucketfs/__init__.py) and the test(s) you can find [here](https://github.com/exasol/bucketfs-python/blob/main/test/integration/bucketfs_test.py).

**Comment by Nicoretti** [comment]
This [PR](https://github.com/exasol/bucketfs-python/pull/93) likely show the important bits best.

**Review by diazona — CHANGES_REQUESTED** [review]
Thanks for your patience. I like the general idea, I just left some suggestions (which I'm happy to discuss if there's anything you don't agree with).

As for commit organization, I would suggest combining the commits that split the certificate/key files and that make the corresponding changes in the code, so you have three commits in all:

- One that fixes the ordering of `cert` and `key`
- One that splits the key into a separate file and makes corresponding updates in the code
- One that adds the `certificate()` method (and `url()` if we wind up keeping that one)

Optionally you could break out the tests into a fourth commit, but that's not necessary. The goal is basically to make sure that each commit has a clear purpose, and that the code compiles on each commit (to make bisection easier, if we ever need to do that). This is not a requirement, though, so it's your call if you want to go that far.

File: tests/test_https.py
Comment (Nicoretti) [reply tests/test_https.py:62]: Nice, that one hadn't been on my radar!

File: tests/test_https.py
Comment (Nicoretti) [reply tests/test_https.py:66]: Actually, writing it this way better highlights the true intent of the test. It's not about ensuring no exceptions are raised. It's specifically about making sure that one particular exception is never raised. That said, I also see the appeal of opting for less code complexity in the tests.

File: pytest_localserver/https.py
Comment (Nicoretti) [reply pytest_localserver/https.py:138]: > I'm actually somewhat opposed to adding this method, because it's quite trivial for client code to construct the URL from `server_address`. I mean, that's basically why `server_address` exists.
> 
> Also, the thought has crossed my mind to add `url_for(path: str)`, like pytest-httpserver does, and that would clash with having a plain old `url()` method.

While I understand your point, I'd like to offer a counter perspective:
For consistency with the `httpserver`, it should have a `url` property, and this property should function as expected without any need for manual intervention. Ideally, we could just inherit the `url` property directly from the `httpserver`. However, it seems this approach is currently unfeasible due to issues with the SAN section of the certificate. I suspect the omission of `127.0.0.1` from the SAN was an oversight, given its relevance to the use cases for which the certificate is intended. Therefore, in the long run, regenerating the certificate to include this might make sense.

File: pytest_localserver/https.py
Comment (Nicoretti) [reply pytest_localserver/https.py:133]: Honestly, I believe that not having to create the certificate oneself is a major advantage compared to `pytest-httpserver`. I think reissuing the certificate in the future would be a better approach.

Still it's your call as maintainer.

**Comment by Nicoretti** [comment]
> ...
> Optionally you could break out the tests into a fourth commit, but that's not necessary. The goal is basically to make sure that each commit has a clear purpose, and that the code compiles on each commit (to make bisection easier, if we ever need to do that). This is not a requirement, though, so it's your call if you want to go that far.

I totally understand the appeal of having clean and "building" commits (I originally come from a Gerrit-based workflow). While GitHub makes this workflow tricky for various reasons (imho), I'm totally with you on this. So, I suggest once you are satisfied with the PR, I'll squash and rebase it into 3-4 commits.

File: pytest_localserver/https.py
Comment (diazona) [reply pytest_localserver/https.py:133]: Oh interesting, I didn't realize pytest-httpserver doesn't provide a default certificate.

Anyway, I wouldn't mind adding `127.0.0.1` to the SAN field in a separate PR (or I guess it could even be a separate commit in this PR, though I generally prefer splitting different features into different PRs). It's just that since we're not doing that yet, I thought it would be best to make the documentation reflect the current state of things for now.

File: pytest_localserver/https.py
Comment (diazona) [reply pytest_localserver/https.py]: Oh, yes I see what you mean. Somehow I missed that there was already a `url` property inherited from `WSGIServer`. (As you might know, I'm not the original author of most of this code, so I'm not as familiar with it as if it were my own project from the start.)

What would you think about making the default value of the `host` parameter be `localhost` in `WSGIServer` and `ContentServer` as well, for consistency with `SecureContentServer`? I don't think the current API makes any guarantees about specifically which hostname will be used, so changing from `127.0.0.1` to `localhost` would be consistent with the API spec. And in fact the documentation mentions `localhost`... if we just consistently use `localhost` everywhere, that should clear up any conflicts. (Unless @redtoad knows of a reason to use the IP address?)

We can still add `127.0.0.1` to the certificate SAN, but given that it's less common to use raw IP addresses with HTTPS, I think if we're going to standardize on anything as the default hostname, it should be `localhost`.

**Review by diazona — COMMENTED** [review]
Whoops I forgot to "officially" leave a review when I commented

File: pytest_localserver/https.py
Comment (Nicoretti) [reply pytest_localserver/https.py:133]: > Oh interesting, I didn't realize pytest-httpserver doesn't provide a default certificate.
> 

No it doesn't however, integrating it using the suggested method [Running an HTTPS server](https://pytest-httpserver.readthedocs.io/en/latest/howto.html#running-an-https-server) is fairly stright forward. Nonetheless, the `pytest-httpserver` method presents several challenges for my specific application, particularly in testing various scenarios involving self-signed certificates (therefore I have choosen `pytest-localserver`).

> Anyway, I wouldn't mind adding `127.0.0.1` to the SAN field in a separate PR (or I guess it could even be a separate commit in this PR... I personally prefer splitting different features into different PRs, but it's not important enough that I'd really insist on it here). I was just thinking that since we're not doing that yet, I thought it would be best to make the documentation reflect the current state of things for now.
> 
> Honestly, it wouldn't be too bad to have one commit that regenerates the certificate with `127.0.0.1` in the SAN and also splits it and the key into separate files. It is sort of bending the principle of making each commit focus on one thing, but those two changes are closely related enough that it seems like a reasonable exception in my opinion. 🤷 So if you want to do that in this PR, feel free to do so, I'm on board.

After giving this some more thinking, I believe that adding `127.0.0.1` to the SAN (Subject Alternative Name) field would be the most straightforward solution. Here's why:

1. It allows us to utilize the URL component from the parent class directly.
2. The URL will behave as expected, closely mirroring the behavior of the HTTP version.
3. Using `localhost` can lead to certain complications, as outlined in the video [Don't Use localhost (intermediate) - Anthony Explains](https://www.youtube.com/watch?v=98SYTvNw1kw).

So if you on board let's do it this way.

**Comment by diazona** [comment]
No problem at all, we're in no rush here and it's fine if this takes a few days or even longer to wrap up. I'll be around to help get it finished up whenever you're ready!

**Comment by Nicoretti** [comment]
@diazona just updated the PR let me know if you need me to do further changes.

File: pytest_localserver/https.py
Comment (diazona) [reply pytest_localserver/https.py]: Sounds good, we're agreed on adding `127.0.0.1` to the SAN field :+1:

As for the hostname, I'd rather stick with using `localhost` consistently throughout. The complications mentioned in that video are relevant in fairly niche situations, not the typical use case for this package. But we don't need to make that change as part of this PR. So I don't think any further changes are needed on this point.

**Review by diazona — APPROVED** [review]
Looks good now! Thanks again for the changes.

**Comment by diazona** [comment]
BTW did you have an interest in getting this PR into a release soon? I hadn't planned on making another release until version 1, whenever that happens, but it'll probably be a while until then. If you'd like these changes available on PyPI earlier, I can make a prerelease to include them or even a version 0.9.

**Comment by Nicoretti** [comment]
@diazona, there's no immediate rush on my end. My goal is to enhance coverage for certain edge cases eventually.

**Comment by diazona** [comment]
Sounds good, I'll leave it for whenever the next release happens, then. Thanks again!

**Pull request merged**
